libraries/bot-builder/src/main/java/com/microsoft/bot/builder/CloudAdapterBase.java

```diff
@@ -0,0 +1,456 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.bot.builder;
+
+import com.microsoft.bot.connector.ConnectorClient;
+import com.microsoft.bot.connector.authentication.ClaimsIdentity;
+import com.microsoft.bot.connector.rest.RestOAuthClient;
+import com.microsoft.bot.schema.Activity;
+import com.microsoft.bot.schema.ResourceResponse;
+import com.microsoft.bot.schema.ConversationReference;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.HttpURLConnection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An adapter that implements the Bot Framework Protocol and can be hosted in different cloud environments
+ * both public and private.
+ */
+public abstract class CloudAdapterBase extends BotAdapter {
+
+    protected BotFrameworkAuthentication botFrameworkAuthentication;
+    protected Logger logger = LoggerFactory.getLogger(CloudAdapterBase.class);
+
+    /**
+     * Gets a {@link Logger} to use within this adapter and its subclasses.
+     * @return The {@link Logger} instance for this adapter.
+     */
+    public Logger getLogger() {
+        return logger;
+    }
+
+    /**
+     * Gets the {@link BotFrameworkAuthentication} instance for this adapter.
+     * @return The {@link BotFrameworkAuthentication} instance for this adapter.
+     */
+    public BotFrameworkAuthentication getBotFrameworkAuthentication() {
+        return botFrameworkAuthentication;
+    }
+
+    /**
+     * Initializes a new instance of the {@link CloudAdapterBase} class.
+     * @param withBotFrameworkAuthentication The cloud environment used for validating and creating tokens.
+     */
+    protected CloudAdapterBase(BotFrameworkAuthentication withBotFrameworkAuthentication) {
+        if (withBotFrameworkAuthentication == null) {
+            throw new IllegalArgumentException("withBotFrameworkAuthentication cannot be null");
+        }
+        this.botFrameworkAuthentication = withBotFrameworkAuthentication;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<ResourceResponse[]> sendActivities(TurnContext context, List<Activity> activities) {
+        if (context == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("context"));
+        }
+
+        if (activities == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("activities"));
+        }
+
+        if (activities.size() == 0) {
+            return Async.completeExceptionally(
+                new IllegalArgumentException("Expecting one or more activities, but the array was empty.")
+            );
+        }
+
+        logger.info(String.format("sendActivities for %d activities.", activities.size()));
+
+        return CompletableFuture.supplyAsync(() -> {
```

Comment:
Never seen this supplyAsync thing. Not sure what this is for. It's not necessarily related to the approval of this PR though

Reply:
I took the same implementation from BotFrameworkAdapter as it was the same code. However, if i'm not wrong, it's very similar to the `thenCompose`